### DISCUSSION TITLE
Can't build with esp-idf framework, only the default arduino. Want to use menuconfig and use PSRAM

### DISCUSSION THREAD

**Discussion by jdeltoft** [mrcodetastic/ESP32-HUB75-MatrixPanel-DMA#557] [Q&A] [answered]
I'm loving this project so far, and have things working with 4 64x64 boards chained.  But this uses a big chunk of DRAM that I would like to move to external RAM if possible.  

> #define LED_MATRIX_DATA_SIZE 49152  // 64x64 (pixels per panel) * 4 (num panels) * 3 (r g and b)
> ...
> EXT_RAM_NOINIT_ATTR uint8_t albumArtData[LED_MATRIX_DATA_SIZE+1] = { 0 };

But this didn't move it.  I think I need to enable "CONFIG_SPIRAM_ALLOW_BSS_SEG_EXTERNAL_MEMORY" with menuconfig?

I was looking at these:

* https://github.com/platformio/platformio-core/issues/3434
* # [enable external RAM on ESP32](https://community.platformio.org/t/correct-settings-to-enable-external-ram-on-esp32/18607)
* https://community.platformio.org/t/how-do-i-use-psram-in-esp32-s2-wrover/29408

I'm on mac and I only use pio CLI. I think I need to use the "esp-idf" framework to get the `pio run -t menuconfig` update to work, but this overall project isn't building with "esp-idf". Just the default "arduino".

I see platformio.ini has esp-idf references but not by default.  Is there anyone that knows how to move over to esp-idf?

**Comment by solhuebner** [comment, marked as answer]
@jdeltoft there is a section in the readme about the usage of PSRAM: https://github.com/mrfaptastic/ESP32-HUB75-MatrixPanel-DMA/blob/master/README.md#required-memory

Are you using the mentioned combination of an ESP32-S3 and OCTAL SPI-RAM (PSTRAM) (i.e. ESP32 S3 N8R8 variant)?

**Reply by jdeltoft** [reply]
Thanks, let me dig into that section.  In answering your question, I realized though that I don't believe have PSRAM.  I'm on esp32-wroom-32 which per https://en.wikipedia.org/wiki/ESP32 doesn't have PSRAM.  

<img width="957" alt="image" src="https://github.com/mrfaptastic/ESP32-HUB75-MatrixPanel-DMA/assets/4506303/0c0d02f1-ebbd-47ba-9cd8-4386282e8a7d">

Here is a snapshot of my build:

<img width="1407" alt="image" src="https://github.com/mrfaptastic/ESP32-HUB75-MatrixPanel-DMA/assets/4506303/d5f68d78-df26-463f-801c-64d80d356e72">

The ram there is a bit misleading (to me at first anyway):

```
CONFIGURATION: https://docs.platformio.org/page/boards/espressif32/nodemcu-32s.html
PLATFORM: Espressif 32 (6.5.0+sha.ee1dd5a) > NodeMCU-32S
HARDWARE: ESP32 240MHz, 320KB RAM, 4MB Flash
...
RAM:   [===       ]  28.9% (used 94548 bytes from 327680 bytes)
Flash: [========= ]  87.5% (used 1146241 bytes from 1310720 bytes)
```

Here is my running board printing out deltas it sees in free heap:

<img width="877" alt="image" src="https://github.com/mrfaptastic/ESP32-HUB75-MatrixPanel-DMA/assets/4506303/8c6904ea-d360-4a6c-b60c-917ed92160ab">

**Reply by solhuebner** [reply]
The issue is also that the normal ESP32 would not work even if you have the PSRAM version. Only the ESP32 „S3” (different chip) in the very specific OCTAL SPI-RAM (PSTRAM) (i.e. ESP32 S3 N8R8 variant) should work.

**Reply by jdeltoft** [reply]
Thank you sooo much @solhuebner !  This link above and the reminder for OCTAL SPI-RAM were very helpful.  Found some nice ESP32-S3-WROOM-1 boards with N16R8 (basically pennies more than N8R8 ...)  

<img width="682" alt="image" src="https://github.com/mrfaptastic/ESP32-HUB75-MatrixPanel-DMA/assets/4506303/616ef710-fe49-441f-b76d-43b858cc9405">

Now just need to wait for them to test. Good time to go finally make the PCB for this while I'm waiting.

**Reply by jdeltoft** [reply]
Got the S3s!!  Heap, awesome.  So far though pins not setting up correctly:

https://github.com/mrfaptastic/ESP32-HUB75-MatrixPanel-DMA/discussions/573
